geps/gep-3155/index.md

```diff
@@ -0,0 +1,264 @@
+
+## Goals
+
+* Add sufficient configuration that basic mutual TLS is possible between Gateways and
+  Backends
+* Enable the optional use of SPIFFE for Backend mutual TLS
+
+## Non-Goals
+
+* Define how automatic mTLS should be implemented with Gateway API
+
+## Introduction
+
+This is a wide ranging GEP intending to cover three additions to the API that all
+have a shared goal - enabling backend mutual TLS with Gateway API. Although this
+specific GEP focuses on manual configuration across the board, the hope is that
+it will also enable higher level automation to simplify this process for users.
+
+## API
+
+### Client Certs on Gateways
+
+A key requirement of mutual TLS is that the Gateway can provide a client cert to the
+backend. This adds that configuration to both Gateway and Service (via 
+BackendTLSPolicy).
+
+#### Gateway-level (Core support)
+Specifying credentials at the gateway level is the default operation mode, where all
+backends will be presented with a single gateway certificate.
+
+#### Service-level (Implementation-specific support)
+Specifying credentials at the service level allows to provide different credentials
+to connect to specific backends or completely disable client certificates. This allows
+to support a wide variety of corner cases, mostly around application-compatibility, e.g.:
+- given OOTB software requires clients to use self-signed, certificates issued by a
+  different issuer, or certificates with a very specific SAN.
+- given OOTB software uses optional certificates for end-user authentication, if the
+  certificate is missing, it falls back to a different application-level authentication
+  method - in such case not providing any certificate is the only way to operate
+
+**1. Add a new `BackendTLS` field at the top level of Gateways**
+
+```go
+type GatewaySpec struct {
+  // BackendTLS configures TLS settings for when this Gateway is connecting to
+  // backends with TLS.
+  BackendTLS GatewayBackendTLS `json:"backendTLS,omitempty"'
+}
+
+type GatewayBackendTLS struct {
+  // ClientCertificateRef is a reference to an object that contains a Client
+  // Certificate.
+  //
+  // References to a resource in different namespace are invalid UNLESS there
+  // is a ReferenceGrant in the target namespace that allows the certificate
+  // to be attached. If a ReferenceGrant does not allow this reference, the
+  // "ResolvedRefs" condition MUST be set to False for this listener with the
+  // "RefNotPermitted" reason.
+  //
+  // ClientCertificateRef can reference to standard Kubernetes resources, i.e.
+  // Secret, or implementation-specific custom resources.
+  //
+  // This setting can be overriden on the service level by use of BackendTLSPolicy.
+  ClientCertificateRef SecretObjectReference `json:"clientCertificateRef,omitempty"`
+}
+```
+
+**2. Add a new `ClientCertificateMode` and `ClientCertificateRef` fields at the top
+level of BackendTLSPolicy**
```

Comment:
I don't know the best way to word this, but it really feels like we need to tell people that you shouldn't use this without a good reason. Because otherwise people will look at the spec of `BackendTLSPolicy` and assume that this is the right place to configure client certificate details.

I also still think that having this config makes the client certificate used here basically worthless for actual authentication - it is essentially `curl -k` for Gateways.

As it is, I suspect that most examples we see will use this and not the Gateway level config, because this is easily within the control of the service owner, and requires no interaction with the Gateway owner.

These feelings aside, the actionable feedback here is to have some sort of callout that Client Certificate config should preferably be in the Gateway stanza, not this one. This should only be used when the Gateway option doesn't work out for some reason.